### PR TITLE
Adding active particle docs

### PR DESCRIPTION
**Pull request by fearmayo**
I'v e started to add/update the AP framework sections.

### CODE REVIEW

File: doc/manual/source/physics/active_particles.rst [hunk 1]
```diff
@@ -0,0 +1,137 @@
+.. _active_particles:
+
+
+Active Particles
+============================================
+
+Active Particles (APs) were introduced into Enzo in version 2.6. They were originally part of the
+of the Enzo-3.0 effort and were developed by a number of authors including Nathan Goldbaum,
+Matt Turk, John Wise, John Regan, Oliver Hahn and others. Their design philosophy was that the
```
Comment (bwoshea) [inline doc/manual/source/physics/active_particles.rst]: please add Brian Crosby and Greg Meece to this list!  They contributed quite a few bug fixes back to Nathan.

File: doc/manual/source/physics/active_particles.rst [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+The ``Skeleton`` particle is simply an example particle heavily commented so allow a user/developer to develop their
+own AP. All the other particles with the exception of the ``AccretingParticle`` and the ``SmartStar`` particle were ported from the
+Star Object particle implementation described in :ref:`star_particles`.
+
+
+Current Limitations of Active Particles
+_______________________________________
+
+Most of the AP types have not been tested - though in principle do work correctly. Caution and some testing of the particle types is advisable for now.
+Multiple AP types can not currently be run together.
```
Comment (bwoshea) [inline doc/manual/source/physics/active_particles.rst]: can this be put in bold?

Comment (fearmayo) [reply doc/manual/source/physics/active_particles.rst]: Sure

File: doc/manual/source/physics/active_particles.rst (hunk 1 above)
Comment (fearmayo) [reply doc/manual/source/physics/active_particles.rst]: Good call. I'll update.

File: doc/manual/source/physics/active_particles.rst [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+``SmartStarFeedbackJetsThresholdMass`` (external)
+    Jets are only ejected once this amount of mass is available for ejected after an accretion event. Therefore, if there is very limited
+    accretion and this parameter is set high then jets will be very infrequent. In units of solar masses. Default: 1.0
+
+``SmartStarSuperEddingtonAdjustment`` (external)
+    As accretion rates exceed the canonical Eddington rate the radiative efficiency of the feedback changes. We use the fits from Madau et al. (https://arxiv.org/pdf/1402.6995.pdf)
+    to adjust the efficiency when accretion enters the super-critical regime. The fits are based on the slim-disk model of accretion which generate inefficient feedback.
+    Default: 1
+
```
Comment (kohdaegene) [inline doc/manual/source/physics/active_particles.rst]: I think either this list of particle descriptions should either be copied in the parameter list, or moved there and linked from here?

Comment (fearmayo) [reply doc/manual/source/physics/active_particles.rst]: Hi DK, 
What did you mean here? The bit of text that is referred to here doesn't exactly match your comment.

Comment (kohdaegene) [reply doc/manual/source/physics/active_particles.rst]: i meant this entire section lines 74-140.

File: doc/manual/source/physics/active_particles.rst [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+The ``Skeleton`` particle is simply an example particle heavily commented so allow a user/developer to develop their
+own AP. All the other particles with the exception of the ``AccretingParticle`` and the ``SmartStar`` particle were ported from the
+Star Object particle implementation described in :ref:`star_particles`.
+
+
+Current Limitations of Active Particles
+_______________________________________
+
+Most of the AP types have not been tested - though in principle do work correctly. Caution and some testing of the particle types is advisable for now.
```
Comment (DeovratPrasad) [inline doc/manual/source/physics/active_particles.rst]: I tested 'Cen-Ostriker' Active Particle. It seems to be doing fine.

File: doc/manual/source/physics/active_particles.rst [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+where ``<YourActiveParticleType>`` is one of the currently available AP types. 
+
+Types of Active Particles
+_________________________
+
+At the moment there are a number of AP type implementations. However, with the exception of the "SmartStar"
+AP type none of the implementations have been (robustly) tested.
+
+* SmartStar
+* Accreting Particle
```
Comment (DeovratPrasad) [inline doc/manual/source/physics/active_particles.rst]: This is a comment related to the Enzo code file not the documentation. In the file 'ActiveParticle_AccretingParticle.h' and 'ActiveParticle_SmartStar.h' there is a class  'DepositMass' defined. One of the argument for this class is 'GalaxyParticleID' in both which I guess is an error getting carried over due to using the modified  'ActiveParticle_GalaxyParticle.h' for these two cases.

File: doc/manual/source/physics/active_particles.rst [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+
+Current Limitations of Active Particles
+_______________________________________
+
+Most of the AP types have not been tested - though in principle do work correctly. Caution and some testing of the particle
+types is advisable for now.
+
+**Multiple AP types can not currently be run together.** This isn't a fundamental limitation. In principle multiple APs can work
+together without difficulty. Some communication work needs to be undertaken to make this work.
```
Comment (DeovratPrasad) [inline doc/manual/source/physics/active_particles.rst]: (Y)

Comment (DeovratPrasad) [inline doc/manual/source/physics/active_particles.rst]: I see the issue primarily related to creation of the buffer list of particles with the Particle Type ID.

File: doc/manual/source/physics/active_particles.rst (hunk 5 above)
Comment (fearmayo) [reply doc/manual/source/physics/active_particles.rst]: OK. That should be a separate PR though. Can you raise an issue on it or open a PR?

File: doc/manual/source/physics/active_particles.rst (hunk 6 above)
Comment (fearmayo) [reply doc/manual/source/physics/active_particles.rst]: Probably.

File: doc/manual/source/physics/active_particles.rst (hunk 3 above)
Comment (fearmayo) [reply doc/manual/source/physics/active_particles.rst]: OK I see what you mean. I'll look into that. Thanks!

File: doc/manual/source/physics/active_particles.rst (hunk 4 above)
Comment (fearmayo) [reply doc/manual/source/physics/active_particles.rst]: Great! I'll update.

File: doc/manual/source/physics/active_particles.rst (hunk 3 above)
Comment (fearmayo) [reply doc/manual/source/physics/active_particles.rst]: Thanks DK that's done now.

**Comment by kohdaegene** [comment]
i realized that there might be some minimal conflicts with PR#83 which was just merged, where i had created some filler headings for active particles. other than that, this looks good to go. this PR should take precedence

**Comment by bwoshea** [comment]
Looks good to me!  We need a third approve here.

File: doc/manual/source/parameters/index.rst [hunk 7]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+
+To allow for the creation of an active particle the following line must be added to the parameter file or restart file:
+``AppendActiveParticleType = <ActiveParticleType``
+See :ref:`active_particles` for the types of active particles that exist.
+
+SmartStar Parameters
+~~~~~~~~~~~~~~~~~~~~
+
+SmartStar Accretion
+^^^^^^^^^^^^^^^^^^^^^^
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
^^^^^^^^^^^^^^^^^^^
```

File: doc/manual/source/parameters/index.rst [hunk 8]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+
+Default: 8
+
+
+Accretion is never capped by default but can capped at the Eddington rate as notes below. 
+   
+
+
+SmartStar Feedback
+^^^^^^^^^^^^^^^^^^^^^^
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
^^^^^^^^^^^^^^^^^^
```

File: doc/manual/source/parameters/index.rst [hunk 9]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+
+Set to 8 to employ the converging mass flux approach. This is not based on any indirect properties of the
+gas as above and simply measures the flow of gas through the accetion radius of the ``SmartStar``. It is based
+on the accretion mechanism used by Bleuler in Ramses https://arxiv.org/pdf/1409.6528.pdf and in the Enzo simulations
+by Regan et al. (2018, 2019) https://arxiv.org/pdf/1803.04527.pdf, https://arxiv.org/pdf/1811.04953.pdf
+
+Default: 8
+
+
+Accretion is never capped by default but can capped at the Eddington rate as notes below.
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
Accretion is never capped by default but can be capped at the Eddington rate as notes below. 
```

File: doc/manual/source/physics/star_particles.rst [hunk 10]
```diff
@@ -1,7 +1,7 @@
 .. _star_particles:
 
 
-Star, Black Hole, Sink, and Active Particles
+Star, Black Hole and Sink Particles
 ============================================
```
Comment (brittonsmith) [inline doc/manual/source/physics/star_particles.rst]:
```suggestion
===================================
```

File: doc/manual/source/parameters/index.rst [hunk 11]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
 
+
+
+.. _active_particles_parameters:
+
+Active Particles
+~~~~~~~~~~~~~~~~
+
+To allow for the creation of an active particle the following line must be added to the parameter file or restart file:
+``AppendActiveParticleType = <ActiveParticleType``
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
``AppendActiveParticleType = <ActiveParticleType>``
```

File: doc/manual/source/parameters/index.rst [hunk 12]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+SmartStar Parameters
+~~~~~~~~~~~~~~~~~~~~
+
+SmartStar Accretion
+^^^^^^^^^^^^^^^^^^^^^^
+
+``SmartStarAccretion`` (external)
+
+Set to 1 to turn on spherical Bondi-Hoyle accretion based on the formalisms
+presented in Krumholz et al. (2004) - https://iopscience.iop.org/article/10.1086/421935/pdf
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
presented in `Krumholz et al. (2004) <https://iopscience.iop.org/article/10.1086/421935/pdf>`__
```

File: doc/manual/source/parameters/index.rst [hunk 13]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+
+Set to 4 to turn on angular momentum limited accretion based on the formalism adopted
+by Rosas-Guevara et al. (2015) - https://arxiv.org/pdf/1312.0598.pdf
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
by `Rosas-Guevara et al. (2015) <https://arxiv.org/pdf/1312.0598.pdf>`__.
```

File: doc/manual/source/parameters/index.rst [hunk 14]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+
+Set to 5 to again turn on spherical Bondi-Hoyle accretion but this time the
+formalism includes a correction which accounts for the vorticity of the gas
+based on Krumholz et al. (2005) - https://iopscience.iop.org/article/10.1086/426051/pdf
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
based on `Krumholz et al. (2005) <https://iopscience.iop.org/article/10.1086/426051/pdf>`__.
```

File: doc/manual/source/parameters/index.rst [hunk 15]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+
+Set to 6 to use the viscous angular momentum prescription from DeBuhr et al. (2010)
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
Set to 6 to use the viscous angular momentum prescription from `DeBuhr et al. (2010) <https://arxiv.org/pdf/0909.2872.pdf>`__.
```

File: doc/manual/source/parameters/index.rst [hunk 16]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+- https://arxiv.org/pdf/0909.2872.pdf
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
```

File: doc/manual/source/parameters/index.rst [hunk 17]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+  
+Set to 7 to use a  modified accretion rate based on the alpha-disk model of Shakura & Sunyaev (1973)
+This scheme is based on Cen et al. https://iopscience.iop.org/article/10.1088/0004-637X/755/1/28/pdf
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
This scheme is based on `Cen et al. <https://iopscience.iop.org/article/10.1088/0004-637X/755/1/28/pdf>`__.
```

File: doc/manual/source/parameters/index.rst [hunk 18]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+
+Set to 8 to employ the converging mass flux approach. This is not based on any indirect properties of the
+gas as above and simply measures the flow of gas through the accetion radius of the ``SmartStar``. It is based
+on the accretion mechanism used by Bleuler in Ramses https://arxiv.org/pdf/1409.6528.pdf and in the Enzo simulations
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
on the accretion mechanism used by `Bleuler <https://arxiv.org/pdf/1409.6528.pdf>`__  in Ramses and in Enzo
```

File: doc/manual/source/parameters/index.rst [hunk 19]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+by Regan et al. (2018, 2019) https://arxiv.org/pdf/1803.04527.pdf, https://arxiv.org/pdf/1811.04953.pdf
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
by Regan et al. (`2018 <https://arxiv.org/pdf/1803.04527.pdf>`__, `2019 <https://arxiv.org/pdf/1811.04953.pdf>`__).
```

File: doc/manual/source/parameters/index.rst [hunk 20]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+    Default: 0
+
+``SmartStarBHFeedback`` (external)
+    This is a master switch on black hole feedback. Must be turned on if you want black hole feedback. Default: 0
+
+``SmartStarBHRadiativeFeedback`` (external)
+    This parameter controls whether black hole radiative feedback gets turned on or not. When turned on the radiative
+    feedback from a black hole depends both on the mass of the black hole and the accretion rate onto the black hole. Both of these
+    quantities are captured and stored as part of the ``SmartStar``. Details of the SED used can be found in the appendix of
+    https://arxiv.org/abs/1811.04953 and is made from assuming a multi-colour disk for the accretion disk and a corona fit by a
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
    `Regan et al. 2019 <https://arxiv.org/abs/1811.04953>`__ and is made from assuming a multi-colour disk for the accretion disk and a corona fit by a
```

File: doc/manual/source/parameters/index.rst [hunk 21]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+    power law.  The radiation emitted by the accretion disk is hard-coded is be emitted by 5 bins with energies of
+    2.0 eV, 12.8 eV, 19.1 eV, 217.3 eV and 5190 eV. The fraction of energy assigned to each bin is then determined by the mass of the
+    black hole and the associated accretion rate at a given time. The formulism is valid for black hole masses between 1 Msolar and
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
    black hole and the associated accretion rate at a given time. The formalism is valid for black hole masses between 1 Msolar and
```

File: doc/manual/source/parameters/index.rst [hunk 22]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+
+``SmartStarBHThermalFeedback`` (external)
+    This parameter controls whether the black hole thermal feedback gets turned. Thre thermal energy is generated by feedback through the
+    accretion process. If this is turned on then the ``SmartStarBHRadiativeFeedback`` should presumably be turned off unless you have a
+    good reason to include both. The efficiency, epsilon, depends on both the spin of the black hole and the ISCO oribit. In order to
+    calculate this accurately Eqn 32 from Abromowicz & Fragile (2013) is used. See ActiveParticle_SmartStar.h. The feedback is released
+    iostropically in a sphere surrounding the SmartStar particle.  Default: 0
+
+``SmartStarBHJetFeedback`` (external)
+    The methodology for this algorithm is based on that of Kim et. al (2011) (https://arxiv.org/pdf/1106.4007.pdf). Jets can be activiated
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
    The methodology for this algorithm is based on that of `Kim et. al (2011) <https://arxiv.org/pdf/1106.4007.pdf>`__. Jets can be activated
```

File: doc/manual/source/parameters/index.rst [hunk 23]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+    when a spinning black hole is accreting. The jets are bipolar and are set along the angular momentum vector of the SmartStar. The
+    velocity of the jet(s) is set by a separate parameter below. No other parameters need to be set to activate the jet.  Default: 0
+
+``SmartStarEddingtonCap`` (external)
+    This parameter allows for accretion onto the SmartStar to be capped at the Eddington limit. I see no good physical reason for doing this in
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
    This parameter allows for accretion onto the SmartStar to be capped at the Eddington limit.
```

File: doc/manual/source/parameters/index.rst [hunk 24]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+    general.  Default: 0
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
    Default: 0
```

File: doc/manual/source/parameters/index.rst [hunk 25]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+
+``SmartStarSpin`` (external)
+    The dimensionless spin of the SmartStar particle. This is a very uncontrained parameter and cannot be readily computed on the fly. This parameter
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
    The dimensionless spin of the SmartStar particle. This is a very unconstrained parameter and cannot be readily computed on the fly. This parameter
```

File: doc/manual/source/parameters/index.rst [hunk 26]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+    should be set if you want to have jet feedback. Setting this is zero and turning on jet feedback wouldn't make sense. The default is set to be 0.7 and
+    this is probably reasonable. 
+    Default: 0.7
+
+``SmartStarSMSLifetime`` (external)
+    This is the lifetime for a supermassive star. After this time has elapsed a SmartStar particle which is behaving like a SMS will collapase
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
    This is the lifetime for a supermassive star in years. After this time has elapsed a SmartStar particle which is behaving like a SMS will collapse
```

File: doc/manual/source/parameters/index.rst [hunk 27]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+    directly into a black hole with no supernova event. Default: 1e6
+
+``SmartStarJetVelocity`` (external)
+    The velocity that the jets are ejected at. Typically jets are observed to travel at a substantial fraction of the speed of light -
+    especially those ejected during periods of high accretion. However,
+    as mass gets entrained on the jet it slows down. The units of this parameter are as a fraction of the speed of light. Default: 0.1
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
    as mass gets entrained in the jet it slows down. The units of this parameter are as a fraction of the speed of light. Default: 0.1
```

File: doc/manual/source/parameters/index.rst [hunk 28]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+
+``SmartStarFeedbackJetsThresholdMass`` (external)
+    Jets are only ejected once this amount of mass is available for ejected after an accretion event. Therefore, if there is very limited
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
    Jets are only ejected once this amount of mass is available for ejection after an accretion event. Therefore, if there is very limited
```

File: doc/manual/source/parameters/index.rst [hunk 29]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+    accretion and this parameter is set high then jets will be very infrequent. In units of solar masses. Default: 1.0
+
+``SmartStarSuperEddingtonAdjustment`` (external)
+    As accretion rates exceed the canonical Eddington rate the radiative efficiency of the feedback changes. We use the fits from Madau et al. (https://arxiv.org/pdf/1402.6995.pdf)
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
    As accretion rates exceed the canonical Eddington rate the radiative efficiency of the feedback changes. We use the fits from `Madau et al. <https://arxiv.org/pdf/1402.6995.pdf>`__
```

File: doc/manual/source/parameters/index.rst [hunk 30]
```diff
@@ -2447,6 +2439,124 @@ The parameters below are currently considered in ``StarParticleCreation`` method
...
+    to adjust the efficiency when accretion enters the super-critical regime. The fits are based on the slim-disk model of accretion which generate inefficient feedback.
```
Comment (brittonsmith) [inline doc/manual/source/parameters/index.rst]:
```suggestion
    to adjust the efficiency when accretion enters the super-critical regime. The fits are based on the slim-disk model of accretion which generates inefficient feedback.
```

File: doc/manual/source/physics/active_particles.rst [hunk 31]
```diff
@@ -0,0 +1,74 @@
+.. _active_particles:
+
+
+Active Particles
+============================================
```
Comment (brittonsmith) [inline doc/manual/source/physics/active_particles.rst]:
```suggestion
================
```

File: doc/manual/source/physics/active_particles.rst [hunk 32]
```diff
@@ -0,0 +1,74 @@
...
+* Accreting Particle
+* Cen-Ostriker
+* GalaxyParticle
+* Kravtsov
+* PopIII
+* RadiationParticle
+* SpringelHernquist
+* Skeleton
+
+The ``Skeleton`` particle is simply an example particle heavily commented so allow a user/developer to develop their
```
Comment (brittonsmith) [inline doc/manual/source/physics/active_particles.rst]:
```suggestion
The ``Skeleton`` particle is simply an example particle heavily commented to allow a user/developer to develop their
```

**Comment by jwise77** [comment]
Hi @fearmayo, can you review and accept @brittonsmith's suggested changes?  Britton can approve when he's ready, and then I'll merge. This is the last substantial PR before the release!

**Comment by fearmayo** [comment]
> Hi @fearmayo, can you review and accept @brittonsmith's suggested changes? Britton can approve when he's ready, and then I'll merge. This is the last substantial PR before the release!

I think that should be done now. I accepted the changes and I can see the commit.

**Pull request merged**